dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.file.dora;
+
+import alluxio.Constants;
+import alluxio.client.block.BlockWorkerInfo;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+/**
+ * An impl of WorkerLocationPolicy.
+ *
+ * Requirements:
+ * 1. Multiple policies available
+ * 2. A policy is stateless Strategy object, which can be a process-unique singleton
+ * 3. Created with factory methods
+ */
+public class ConsistentHashPolicy implements WorkerLocationPolicy {
+  private static final ConsistentHashProvider HASH_PROVIDER =
+      new ConsistentHashProvider(100, Constants.SECOND_MS);
+  /**
+   * This is the number of virtual nodes in the consistent hashing algorithm.
+   * In a consistent hashing algorithm, on membership changes, some virtual nodes are
+   * re-distributed instead of rebuilding the whole hash table.
+   * This guarantees the hash table is changed only in a minimal.
+   * In order to achieve that, the number of virtual nodes should be X times the physical nodes
+   * in the cluster, where X is a balance between redistribution granularity and size.
+   */
+  private final int mNumVirtualNodes;
+  private final AlluxioConfiguration mConf;
+
+  /**
+   * Constructs a new {@link ConsistentHashPolicy}.
+   */
+  public ConsistentHashPolicy(AlluxioConfiguration conf) {
+    mConf = conf;
+    mNumVirtualNodes = conf.getInt(PropertyKey.USER_CONSISTENT_HASH_VIRTUAL_NODE_COUNT);
+  }
+
+  @Override
+  public List<BlockWorkerInfo> getPreferredWorkers(
```

Comment:
I think passing in a worker list here is ambiguous - what is this list and what if the list does not match with the list maintained by the HASH_PROVIDER?

However, a counter-argument is:  The policy object should be a [Strategy object](https://en.wikipedia.org/wiki/Strategy_pattern) And this strategy object should be a singleton and should be stateless. Therefore the strategy object should get the list from external callers.

Can we achieve stateless without needing a worker list argument? I think the Policy can rely on a singleton HASH_PROVIDER. The HASH_PROVIDER relies on another singleton WorkerListUpdater if necessary. All Policy instances will rely on the singletons for the single source of truth worker list.

Reply:
The policy itself should not hold on to a stateful hash provider. It's input should always be provided by the caller. However, the input does not have to be a list of workers, otherwise based on which the policy has to build a hash ring every time it is consulted. The input could be an opaque `WorkerClusterView` object, which carries an immutable, embedded hash ring. The view object is itself provided and updated by the cluster membership service, and fetched by the `FileSystemConext` every time the application does IO.

The hierarchy for the classes looks like

```
Membership service
| generates or updates
v
Worker list
| creates
v
WorkerClusterView
| specializes to in case CH is enabled
v
ConsistenHashBasedView (embeds an immutable hash ring)
| fetched by
v
FileSystemContext
| queried by
v
ConsistentHashPolicy
```

If we add some policy other than consistent hash in the future, the hierarchy will look like:

```
Membership service
| generates or updates
v
Worker list
| creates
v
WorkerClusterView
| specializes to
v
StaticListBasedView (embeds a simple worker list)
| fetched by
v
FileSystemContext
| queried by
v
RoundRobinPolicy
```